### ISSUE TITLE
Get type inference to work with Type.Enum

### ISSUE THREAD

**Issue by joshmossas** [sinclairzx81/typebox#549]
Currently when creating an enum typescript will only infer the values as a `string` or `number`. Below you can see that the `role` variable should have an error but it doesn't

![image](https://github.com/sinclairzx81/typebox/assets/31548851/d8c4720f-efbe-4767-a532-167e50fa65bb)
![image](https://github.com/sinclairzx81/typebox/assets/31548851/ec71ba3d-7123-4f20-90ce-6954392dd86a)

I decided to play around with the Typebox source and found with a slight tweaks to the generic type we can actually get typescript to infer the types properly. 

![image](https://github.com/sinclairzx81/typebox/assets/31548851/77e987eb-cf71-4fd7-8b99-540755918cfc)
![image](https://github.com/sinclairzx81/typebox/assets/31548851/73d46cab-160c-4697-a95f-9676a23cc41e)

I have a fork with a working version of this here https://github.com/joshmossas/typebox
Let me know if you would like to merge this change and I can open a PR.

Cheers

**Comment by sinclairzx81** [comment]
@joshmossas Hi!

Hey, nice work! I actually really like the implementation there, I'd be happy to accept a PR for this :)

Just for a bit of history though. I did have some plans to update Enum to use [Const Generics](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#const-type-parameters) in TS 5.0 (when TB drops support for 4.0), but I think I like your implementation more (mostly due to backwards compatibility with TS 4.0)

For reference, the following would be the generic signatures for Current, Proposed and Future revisions to Enum.

### Current
TS 4.0: Requires caller to specify 'as const' on the passed object literal
```typescript
public Enum<T extends Record<string, string | number>>(item: T, options: SchemaOptions = {}): TEnum<T>
```
### Proposed
TS 4.0: Derive V from item argument
```typescript
public Enum<V extends string | number, T extends Record<string, V>>(item: T, options: SchemaOptions = {}): TEnum<T>
```
### Future
TS 5.0: Auto infer T as const
```typescript
public Enum<const T extends Record<string, string | number>>(item: T, options: SchemaOptions = {}): TEnum<T>
         // ^ added
```
I guess the only reservation I'd have implementing the Proposed is that it adds +1 additional generic arguments (where Const Generics would retain the same generic signature, but at a loss to backwards compatibility with TS 4.0), so weighing these factors, I think your implementation is the way to go (for both TS 4.0 & 5.0), but may revise once there's TS sees significant (or general) adoption of 5.0.

Would be super happy to receive a PR for this. But would it be possible to rename the generic argument from `K` to `V` ? (as the argument relates to the value, not the key)

Many Thanks, this update looks to be a good one! :)
S

**Comment by joshmossas** [comment]
Oh interesting! I wasn't familiar with the const type generic but it looks super useful I'll have to remember that in the future.

I went ahead and made the pull request with the requested changes (renaming `K` to `V`). Funnily enough I had initially named the parameter `V` during my experiments (inside the node_modules folder haha) but somehow I mixed it up when making a fork

**Comment by sinclairzx81** [comment]
@joshmossas All good :) Updates are published on `0.31.4` 

Hey, thanks again for the PR! This update is actually pretty great as quite a few users have asked me for better support for the object literal enum pattern. Having both the inference work without `as const` and TS 4.0 (and future 5.0) compatibility, that's just excellent! :)

All the best!
S
